tools/shared.py

```diff
@@ -2555,19 +2555,29 @@ def add_to_path(dirname):
       logger.debug('closure compiler: ' + ' '.join(cmd))
 
       proc = run_process(cmd, stderr=PIPE, check=False, env=env)
-      if proc.returncode != 0:
-        sys.stderr.write(proc.stderr)
-        hint = ''
-        if not pretty:
-          hint = ' the error message may be clearer with -g1'
-        exit_with_error('closure compiler failed (rc: %d.%s)', proc.returncode, hint)
 
       # XXX Closure bug: if Closure is invoked with --create_source_map, Closure should create a
       # outfile.map source map file (https://github.com/google/closure-compiler/wiki/Source-Maps)
       # But it looks like it creates such files on Linux(?) even without setting that command line
       # flag (and currently we don't), so delete the produced source map file to not leak files in
       # temp directory.
       try_delete(outfile + '.map')
+
+      if len(proc.stderr.strip()) > 0:
+        logger.debug(open(filename, 'r').read())
```

Comment:
Printing the entire intput file to stdout seems a little excessive, even when EMCC_DEBUG=1.

Reply:
It might feel like that, but I think it is warranted.

Every single time when I need to debug Closure (as an Emscripten user), I end up adding a `logger.fatal(open(filename, 'r').read())` here so that I can see what the input file was that was fed to Closure. The errors&warnings that are emitted are always (not overstating here) useless without being able to see the whole context, and it happens often.

Currently to dig up the errors/warnings, you need to set EMCC_DEBUG=1 environment variable, and EM_SAVE_DIR=1 environment variable, and then rerun, figure out which of the captured temp files is the closest one to be the input for closure (we don't actually currently even record one that is the right immediate input, but there is one that is from a few intermediate steps before), and then open that in text editor. Then don't forget to delete the generated temp files to not leave litter around.

We don't document the above workflow. I don't actually know what other people are doing when they get Closure errors, but currently I imagine they are doing something quite creative as well, or try to guess the context from the oneliner Closure prints.

I could not come up with a better way than to logger.debug() the whole input file when one gets errors, and add instructions to how to enable it. Using logger.debug() because the entire input file can be quite overwhelming to print to console.

Reply:
OK, for now that sounds fine.  In the long run maybe we can find a better way.   Ideally when any command fails all its inputs would be available for inspection (so one could re-run the failing command again).  Or perhaps we advertise some kind of `--save-temps` flag that can be used when re-running the failing emcc command.